src/systems/optical_tactile_plugin/OpticalTactilePlugin.cc

```diff
@@ -0,0 +1,682 @@
+          _ecm.Component<components::Pose>(
+            this->dataPtr->model.Entity())->Data();
+
+      // Depth Camera data is float, so convert Pose3d to Pose3f
+      this->dataPtr->sensorWorldPose = ignition::math::Pose3f(
+          sensorPose.Pos().X(), sensorPose.Pos().Y(), sensorPose.Pos().Z(),
+          sensorPose.Rot().W(), sensorPose.Rot().X(), sensorPose.Rot().Y(),
+          sensorPose.Rot().Z());
+    }
+}
+
+//////////////////////////////////////////////////
+void OpticalTactilePlugin::PostUpdate(
+    const ignition::gazebo::UpdateInfo &_info,
+    const ignition::gazebo::EntityComponentManager &)
+{
+    IGN_PROFILE("TouchPlugin::PostUpdate");
+
+    // Nothing left to do if paused.
+    if (_info.paused)
+      return;
+}
+
+//////////////////////////////////////////////////
+void OpticalTactilePluginPrivate::DepthCameraCallback(
+        const ignition::msgs::PointCloudPacked &_msg)
+{
+  // This check avoids running the callback at t=0 and getting
+  // unexpected markers in the scene
+  if (!this->initialized)
+      return;
+
+  this->cameraMsg = _msg;
+  this->newCameraMsg = true;
+}
+
+//////////////////////////////////////////////////
+void OpticalTactilePluginPrivate::UnpackPointCloudMsg(
+                        const ignition::msgs::PointCloudPacked &_msg)
+{
+  // 3D vector to store XYZ points
+  this->imageXYZ = std::vector<std::vector<std::vector<float>>>
+      (_msg.width(),
+      std::vector<std::vector<float>>(_msg.height(),
+      std::vector<float>(3)));
+
+  // Unpack point cloud message
+  std::string msgBuffer = _msg.data();
+  char *msgBufferIndex = msgBuffer.data();
+
+  for (uint32_t j = 0; j < _msg.height(); ++j)
+  {
+    for (uint32_t i = 0; i < _msg.width(); ++i)
+    {
+      int fieldIndex = 0;
+
+      // x coordinate in pixel (i,j)
+      this->imageXYZ[i][j][0] =  *reinterpret_cast<float *>(
+        msgBufferIndex + _msg.field(fieldIndex++).offset());
+      // y coordinate in pixel (i,j)
+      this->imageXYZ[i][j][1] = *reinterpret_cast<float *>(
+        msgBufferIndex + _msg.field(fieldIndex++).offset());
+      // z coordinate in pixel (i,j)
+      this->imageXYZ[i][j][2] = *reinterpret_cast<float *>(
+        msgBufferIndex + _msg.field(fieldIndex++).offset());
+
+      msgBufferIndex += _msg.point_step();
+    }
+  }
+}
+
+//////////////////////////////////////////////////
+void OpticalTactilePluginPrivate::ComputeNormalForces(
+                    const ignition::msgs::PointCloudPacked &_msg,
+                    const bool _visualizeForces)
+{
+  // 2D vector to store normal force vectors
+  std::vector<std::vector<ignition::math::Vector3f>> normalForces(_msg.width(),
+      std::vector<ignition::math::Vector3f>(_msg.height()));
+
+  // Visualization is downsampled according to plugin parameter <resolution>
+  // if <visualize_forces> is set to true
+  bool visualizeRow = true;
+  int rowCounter = 0;
+
+  bool visualizeColumn = true;
+  int columnCounter = 0;
+
+  int index = 1;
```

Comment:
A more descriptive variable name for the index, since there's a lot of indexing going on here.